homeassistant/components/sensor/__init__.py

```diff
@@ -246,6 +247,22 @@ def device_class(self) -> SensorDeviceClass | None:
             return self.entity_description.device_class
         return None
 
+    @final
+    @property
+    def numeric_state_expected(self) -> bool:
+        """Return true if the sensor must be numeric."""
+        if (
+            self.state_class is not None
+            or self.native_unit_of_measurement is not None
+            or self.native_precision is not None
+        ):
+            return True
+        device_class: SensorDeviceClass | None = None
+        with suppress(ValueError):
+            # Sensors with custom device classes are not considered numeric
+            device_class = SensorDeviceClass(str(self.device_class))
+        return not (device_class is None or device_class in NON_NUMERIC_DEVICE_CLASSES)
```

Comment:
I think this makes the logic more readable:
```suggestion
        return device_class is not None and device_class not in NON_NUMERIC_DEVICE_CLASSES)
```